revocation/internal/crl/crl.go

```diff
@@ -0,0 +1,298 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package crl provides methods for checking the revocation status of a
+// certificate using CRL
+package crl
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/revocation/result"
+)
+
+// oidInvalidityDate is the object identifier for the invalidity date
+// CRL entry extension. (See RFC 5280, Section 5.3.2)
+var oidInvalidityDate = asn1.ObjectIdentifier{2, 5, 29, 24}
+
+// maxCRLSize is the maximum size of CRL in bytes
+const maxCRLSize = 10 << 20 // 10 MiB
+
+// Options specifies values that are needed to check CRL
+type Options struct {
+	// HTTPClient is the HTTP client used to download CRL
+	HTTPClient *http.Client
+
+	// AuthenticSigningTime is the time when the certificate's private key is
+	// used to sign the data.
+	AuthenticSigningTime time.Time
+}
+
+// CertCheckStatus checks the revocation status of a certificate using CRL
+//
+// The function checks the revocation status of the certificate by downloading
+// the CRL from the CRL distribution points specified in the certificate.
+//
+// If the invalidity date extension is present in the CRL entry and SigningTime
+// is not zero, the certificate is considered revoked if the SigningTime is
+// after the invalidity date. (See RFC 5280, Section 5.3.2)
+func CertCheckStatus(ctx context.Context, cert, issuer *x509.Certificate, opts Options) *result.CertRevocationResult {
+	if !Supported(cert) {
+		return &result.CertRevocationResult{
+			Result: result.ResultNonRevokable,
+			CRLResults: []*result.CRLResult{{
+				Error: fmt.Errorf("CRL is not supported in the certificate %s", cert.Subject.CommonName),
```

Comment:
This is because Golang error messages are often concatenated with other strings, and starting with a lowercase letter makes them flow better in such cases. However, `CRL` is an acronym, so even when concatenated with other strings, the capitalized `CRL` still makes sense.